src/chat.rs

```diff
@@ -152,6 +180,130 @@ impl ChatId {
+                            bail!(
+                                "{} is not verified; cannot enable protection.",
+                                contact.get_display_name()
+                            );
+                        }
+                    }
+                }
+                Chattype::Undefined => bail!("set protection: undefined group type"),
+            },
+            ProtectionStatus::Unprotected => {}
+        };
+
+        context
+            .sql
+            .execute(
+                "UPDATE chats SET protected=? WHERE id=?;",
+                paramsv![protect, self],
+            )
+            .await?;
+
+        context.emit_event(EventType::ChatModified(self));
+
+        // make sure, the receivers will get all keys
+        reset_gossiped_timestamp(context, self).await?;
+
+        Ok(())
+    }
+
+    /// Returns a stock message saying that protection status has changed.
+    pub(crate) async fn get_protection_msg(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+        from_id: u32,
+    ) -> String {
+        context
+            .stock_system_msg(
+                match protect {
+                    ProtectionStatus::Protected => StockMessage::ProtectionEnabled,
+                    ProtectionStatus::Unprotected => StockMessage::ProtectionDisabled,
+                },
+                "",
+                "",
+                from_id,
+            )
+            .await
+    }
+
+    /// Send protected status message to the chat.
+    ///
+    /// This sends the message with the protected status change to the chat,
+    /// notifying the user on this device as well as the other users in the chat.
+    /// If `promoted` is false this means the chat only exists on this device so far
+    /// and does not need to be sent out.
+    /// In this case an local info message is added to the chat.
+    pub(crate) async fn add_protection_msg(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+        promoted: bool,
+        from_id: u32,
+    ) -> Result<(), Error> {
+        let msg_text = self.get_protection_msg(context, protect, from_id).await;
+
+        if promoted {
+            let mut msg = Message::default();
+            msg.viewtype = Viewtype::Text;
+            msg.text = Some(msg_text);
+            msg.param.set_cmd(match protect {
+                ProtectionStatus::Protected => SystemMessage::ChatProtectionEnabled,
+                ProtectionStatus::Unprotected => SystemMessage::ChatProtectionDisabled,
+            });
+            send_msg(context, self, &mut msg).await?;
+        } else {
+            add_info_msg(context, self, msg_text).await;
+        }
+
+        Ok(())
+    }
+
+    /// Sets protection and sends or adds a message.
+    pub async fn set_protection(
+        self,
+        context: &Context,
+        protect: ProtectionStatus,
+    ) -> Result<(), Error> {
```

Comment:
i'm going to over-think this function signature, bear with me :)

Some of the errors here are treated as just bad failures: the `ensure!()` call, the `::load_from_db()?` one.  But one failure is expected and this function handles reporting it to the user (`.inner_set_protection()`).  Now the interesting part is that `Err` is returned when this function already handled the error of `.inner_set_protection()`, making it impossible to distinguish the difference for the caller.

So my proposal is that `Err` should mean it's unhandled, and the failure for `.inner_set_protection()` should return `Ok`.  Now the caller might still needs to know if the protection was enabled or not.  The simple way would be have a bool in the `Ok` variant, indicating whether it's verified or not.  The simple way is probably practical enough?

Reply:
> making it impossible to distinguish the difference for the caller.

that's true - but i've checked all calling situations - as far as i see, in all cases, one would just handle both cases in the same way. so instead of only bubbling up with `?` you would have to check for Ok(false) as well.

so, i think, until some caller wants to make some sense from these different return values, we can leave things as Result<(), Error>